clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -660,6 +674,74 @@ public void subscribe(TopicPartition... partitions) {
     }
 
     /**
+     * Subscribes to topics matching specified pattern and uses the consumer's group
+     * management functionality. The pattern matching will be done periodically against topics
+     * existing at the time of check.
+     * <p>
+     * As part of group management, the consumer will keep track of the list of consumers that
+     * belong to a particular group and will trigger a rebalance operation if one of the
+     * following events trigger -
+     * <ul>
+     * <li>Number of partitions change for any of the subscribed list of topics
+     * <li>Topic is created or deleted
+     * <li>An existing member of the consumer group dies
+     * <li>A new member is added to an existing consumer group via the join API
+     * </ul>
+     *
+     * @param pattern Pattern to subscribe to
+     */
+    @Override
+    public void subscribe(Pattern pattern) {
+        acquire();
+        try {
+            log.debug("Subscribed to pattern: {}", pattern);
+            this.subscriptions.subscribe(pattern);
+
+            if (this.patternSubscriptionTask == null) {
+                schedulePatternSubscriptionTask(this.patternSubscriptionIntervalMs);
+            }
+        } finally {
+            release();
+        }
+    }
+
+    private void schedulePatternSubscriptionTask(final long interval) {
+        this.patternSubscriptionTask = new PatternSubscriptionTask(interval);
+        client.schedule(this.patternSubscriptionTask, time.milliseconds() + interval);
+    }
+
+    private boolean matchesSubscribedPattern(String topic) {
+        return this.subscriptions.getSubscribedPattern().matcher(topic).matches();
+    }
+
+    private boolean isAlreadySubscribed(String topic) {
+        return this.subscriptions.subscribedTopics().contains(topic);
+    }
+
+    /**
+     * Unsubscribe from topics subscribed via pattern subscription
+     *
+     * @param pattern Pattern to unsubscribe from, must match the pattern subscribed to
+     */
+    public void unsubscribe(Pattern pattern) {
```

Comment:
Hi @guozhangwang , I had some thoughts along the same line and have posted them on the JIRA. Below is the copy.

"Should we support topic and regex subscription at same time? There are a few special cases that will need attention if we decide to have regex and topic subscription work at same time. Say one subscribed to test* and then later explicitly unsubscribed from test1. test1 will get subscribed the next time regex check is run. This can be handled by maintaining a blacklist for a particular pattern. The point is we can support that, however it will have complicated logic. Wondering what will we gain from this complicated logic though. One can have a pattern to subscribe to individual topic or a collection of topics or both. Having support for topic and regex subscription does not buy a lot. If nobody suggests otherwise I will go with making topic and regex subscription mutually exclusive.
Also, in my current implementation I am only supporting subscribing to a single pattern and not multiple patterns as multiple patterns can be combined into a single pattern. Hope that is fine as well."